New scale test for audit logging

graphcareful opened the pull request:
Introduce a new scale test for audit logging. The test environment leverages the kgo-repeater to generate traffic over a set number of topics so that many unique audit messages are generated, making a sort of almost worst-case scenario event for the system.

This would be a case where auditing doesn't have the oppurtunity to combine like messages. These for example would be produce messages coming from the same topic on the same shard (since partition_id is not part of the audit message payload only topic name).  Therefore this test differes from other scale tests which typically use one topic and a large number of partitions. Instead the audit scale test ensures a topic can only have as many partitions as CPUs exist in the cluster, if more are desired, they must be spread across multiple topics.

This patch leverages a recent change in the kgo-repeater to make the program multi topic aware. https://github.com/redpanda-data/kgo-verifier/pull/47

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

* none

oleiman reviewed (commented):
Looks great to me. Very clear. Several random accumulated comments, but nothing blocking.

graphcareful commented on tests/rptest/services/kgo_repeater_service.py:
```diff
@@ -328,22 +330,53 @@ def group_ready():
+        def tuple_op(binary_op, a, b):
+            """Perform binary_op() across all fields of tuple a and b"""
+            assert len(a) == len(b)
+            return tuple(binary_op(a[i], b[i]) for i in range(0, len(a)))
```
sickk

graphcareful commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,359 @@
+            # No auditing will occur without the global feature flag set to True
+            'audit_enabled':
+            True,
+
+            # The default is 8, but for a scale test its desired to increase the
+            # number of partitions to handle the expected load
+            'audit_log_num_partitions':
+            12,
```
I thought about that, but then when i was looking at the other scale tests they had explicitly modified configs to help with performance at scale.

oleiman replied:
ah, totally. makes sense.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ef3ebb6196ead3d0ef8da9d364a0ca1799c0b429..ab70b0fe3cc3a50ea011fc4dcae1f07dd49c9d8c) 
- fixed a bug with throughput calculation

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/7aaf58bf0cd89608a35cd41ec94b895f4212397c..b08c4ab2b0e8be882cca96f8e8f4ad1a0999ed07)
- Implemented Orens feedback
- Fixed bug in audit log manager w.r.t. probe mem calculation
- Added feature to periodically log audit metrics during scale test

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b08c4ab2b0e8be882cca96f8e8f4ad1a0999ed07..24f7652ba7bcf0e4d87c76afc4e1800b01f54c2c)
- Rebased against dev

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/42733#018c6552-f979-42f8-9a8e-23fae2b4e51a:
```
"rptest.tests.audit_log_test.AuditLogTestKafkaApi.test_management"
```
<!--e35bf47f-2b67-44fe-9b83-d437033b0057-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/42733#018c6563-3af9-4a00-8e94-34faaf85d62c
<!--38df04e4-a79d-41e7-ae2b-4900d1620827-->

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/24f7652ba7bcf0e4d87c76afc4e1800b01f54c2c..39f61077501bad90c7f4265fd866d12564c50f93)
- Rebased dev to resolve conflict

michael-redpanda commented on src/v/security/audit/schemas/application_activity.h:
```diff
@@ -89,6 +89,13 @@ class api_activity final : public ocsf_base_event<api_activity> {
     api_activity_unmapped _unmapped;
 
+    ss::sstring api_info() const final {
+        if (_http_request) {
```
nit: I _think_ our style is to use `.has_value()` rather than relying on the bool operator, but I could be mistaken

michael-redpanda commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,412 @@
+    Hardcoded security parameters for the audit scale tests. SASL will be
+    the authentication method of choice
+    """
+    sasl_credentials = SaslCredentials('rob', '123rob456', 'SCRAM-SHA-256')
```
question: who's rob?

graphcareful replied:
no idea

michael-redpanda commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,412 @@
+        self.p99 = latencies[2]
+
+    def _mbps(self, b: float) -> float:
+        return (b / self.time_elapsed) / (1024 * 1024.0)
```
question: why `1024 * 1024.0` and not `1024.0 * 1024.0`?

graphcareful replied:
took it from elsewhere, effectively should be the same

michael-redpanda commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,412 @@
+
+        # We will send huge numbers of messages, so tune down the log verbosity
+        # as we will be comparing performance to a known good baseline
+        kwargs['log_level'] = 'info'
```
question: Did you want to set auditing facility to `trace`?

graphcareful replied:
it would produce a ton of log messages maybe too many? Not really necessary IMO since we are now periodically printing the metrics stuff , that will help us understand whats happening

graphcareful replied:
will change to debug

michael-redpanda replied:
Ah I see, we're just going to collect on when events cannot be inserted, is that it?

michael-redpanda commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,412 @@
+            f'User:{self.user_creds.username}', ['all'], 'topic', '*')
+        self._super_rpk.sasl_allow_principal(
+            f'User:{self.user_creds.username}', ['all'], 'group', '*')
+        # Attempt to list topics to check if the above worked
+        self._user_rpk.list_topics()
```
nit: This does check that the user is created, but doesn't verify that the user has access to the topics we want them to (e.g. if no ACLs provided for the user, this would return nothing). Not sure if that's a problem or not, just pointing it out.

michael-redpanda commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,412 @@
+            # to me able to assert on other properties of the test run.
+            return make_result_set(t1, repeater)
+
+    # Temporarily marking as ok_to_fail so we can observe in CI how fill
```
nit: s/fill/full

michael-redpanda commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,412 @@
+            return make_result_set(t1, repeater)
+
+    # Temporarily marking as ok_to_fail so we can observe in CI how fill
+    # the audit queues are during the scale test so we can appropirately
```
nit: s/appropirately/appropriately

michael-redpanda commented on tests/rptest/scale_tests/audit_log_test.py:
```diff
@@ -0,0 +1,412 @@
+            TopicSpec(partition_count=partitions_per_topic)
+            for _ in range(0, total_topics)
+        ]
+        if extra_partitions > 0:
+            # If partition_limit doesn't evenly divide into the total number of shards,
+            # create an extra topic with the number of partitions containing the diff
+            difference = partitions_per_topic - extra_partitions
+            topics = topics + [TopicSpec(partition_count=difference)]
```
question: Shouldn't partition count be `extra_partitions`?

Say you have a partition limit of 103 and you have 10 partitions per topic, you'd have 3 partitions left over from `scale.partition_limit % partitions_per_topic`

graphcareful replied:
yeah but i am trying to add `7` to make `110` in this case, but now that i think about it, maybe I shouldn't because the bandwidth estimates are based on `scale.partition_limit` and i should be making exactly that

michael-redpanda reviewed (commented):
just a couple of questions, but looks good

michael-redpanda reviewed (approved):
lgtm

oleiman reviewed (approved):
lgtm
